### PR TITLE
Ensure that integer parsing correctly handles non-zero fractional data

### LINKED ISSUE #94971: short.Parse regression in .NET 8.0 from 7.0
### Description

Methods `Int16.Parse` and `Int16.TryParse` produce different result in .NET 7 and 8, when the input is decimal number. In .NET 7 numbers with decimal point are considered invalid and the methods returns error. However, in .NET 8 the methods return integer number, when input has more than 4 digits.

### Reproduction Steps

Example:
```
short.TryParse("10000.1", NumberStyles.Number, CultureInfo.InvariantCulture, out var number);
```

### Expected behavior

The function should return false in both .NET versions.

### Actual behavior

The function returns true in .NET 8 and false in .NET 7.

### Regression?

Yes, as described above.

### Known Workarounds

I found the issue only in Int16. Int32 works the same in .NET 8 and 7.

### Configuration

dotnet --info

.NET SDK:
 Version:           8.0.100
 Commit:            57efcf1350
 Workload version:  8.0.100-manifests.8d38d0cc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22631
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100\

.NET workloads installed:
 Workload version: 8.0.100-manifests.8d38d0cc
 [maui-windows]
   Installation Source: VS 17.8.34309.116
   Manifest Version:    8.0.3/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.maui\8.0.3\WorkloadManifest.json
   Install Type:              Msi

 [maccatalyst]
   Installation Source: VS 17.8.34309.116
   Manifest Version:    17.0.8478/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.maccatalyst\17.0.8478\WorkloadManifest.json
   Install Type:              Msi

 [ios]
   Installation Source: VS 17.8.34309.116
   Manifest Version:    17.0.8478/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.ios\17.0.8478\WorkloadManifest.json
   Install Type:              Msi

 [android]
   Installation Source: VS 17.8.34309.116
   Manifest Version:    34.0.43/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.android\34.0.43\WorkloadManifest.json
   Install Type:              Msi

Host:
  Version:      8.0.0
  Architecture: x64
  Commit:       5535e31a71

.NET SDKs installed:
  8.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

### Other information

The problem appears only for values 10000 and higher in absolute value.

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves #94971

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @stephentoub.

This is fixing a regression introduced in .NET 8, so we likely want to backport into .NET 8 and .NET 9 RC2

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Int32Tests.cs [hunk 1]
```diff
@@ -663,6 +663,46 @@ public static IEnumerable<object[]> Parse_Invalid_TestData()
             yield return new object[] { "2147483649-", NumberStyles.AllowTrailingSign, null, typeof(OverflowException) };
             yield return new object[] { "(2147483649)", NumberStyles.AllowParentheses, null, typeof(OverflowException) };
             yield return new object[] { "2E10", NumberStyles.AllowExponent, null, typeof(OverflowException) };
+
+            // Test trailing non zeros
+
+            yield return new object[] { "-9223372036854775808.1", NumberStyles.Number, CultureInfo.InvariantCulture, typeof(OverflowException) };
```
Comment (tannergooding) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Int32Tests.cs:669]: Just noting for reviewers that this `Parse_Invalid_TestData` is reused by the other integer types as part of their parsing validation logic.

For example: https://github.com/dotnet/runtime/blob/b4d433380c24889565cc8ea7703db1106c2ad0df/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/SByteTests.cs#L266-L269

and 

https://github.com/dotnet/runtime/blob/b4d433380c24889565cc8ea7703db1106c2ad0df/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Int64Tests.cs#L303-L313

**Review by jeffhandley — APPROVED** [review]

**Pull request merged**